src/client/dfs/dfs_sys.c

```diff
@@ -1399,6 +1399,49 @@ dfs_sys_mkdir(dfs_sys_t *dfs_sys, const char *dir, mode_t mode,
 	return rc;
 }
 
+int
+dfs_sys_mkdir_p(dfs_sys_t *dfs_sys, const char *dir_path, mode_t mode, daos_oclass_id_t cid)
+{
+	int         path_len = strnlen(dir_path, PATH_MAX);
+	char       *_path    = NULL;
+	char       *ptr     = NULL;
+	int         rc       = 0;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (dir_path == NULL)
+		return EINVAL;
+	if (path_len == PATH_MAX)
+		return ENAMETOOLONG;
+
+	D_STRNDUP(_path, dir_path, path_len);
+	if (_path == NULL)
+		return ENOMEM;
+
+	/* iterate through the parent directories and create them if necessary */
+	for (ptr = _path + 1; *ptr != '\0'; ptr++) {
+		if (*ptr != '/')
+			continue;
+
+		/* truncate the string here to create the parent */
+		*ptr = '\0';
+		rc   = dfs_sys_mkdir(dfs_sys, _path, mode, cid);
+		if (rc != 0 && rc != EEXIST)
+			D_GOTO(out_free, rc);
+		/* reset to keep going */
+		*ptr = '/';
+	}
+
+	/* create the final directory */
+	rc = dfs_sys_mkdir(dfs_sys, _path, mode, cid);
+	if (rc == EEXIST)
```

Comment:
i think the main issue why your test fails is this line here. dfs is returning EEXIST saying there is an entry (a file) called whoops and you set that to success. i think this is a bug in your PR

Reply:
But if you've already got a file named `/a/b/whoops` and you do `mkdir(/a/b/whoops)`, it shouldn't succeed, right? Without the existence check that I re-added, running mkdir on a file doesn't fail. Conversely, running regular mkdir on a directory that does exist also doesn't fail. Those both seem pretty broken to me.

I'm guessing that handling this in the conditional insert may be difficult or impossible because at that level it's just objects and there's no notion of file vs. directory, right?

Reply:
So even C `mkdir()` sets errno=EEXIST whether it is an existing file or directory.
But the behavior @mjmac is after is similar to what bash `mkdir -p` does.
Existing file will error
```
$ touch file
$ mkdir -p file
mkdir: cannot create directory ‘file’: File exists
```
Existing dir will not
```
$ mkdir dir
$ mkdir -p dir
```

Reply:
I think `dfs_mkdir` returning EEXIST for both cases makes sense, so then `dfs_sys_mkdir_p` should handle this?

Reply:
> running mkdir on a file doesn't fail.

well it doesn't fail with your function but dfs does fail it. you are just converting the error to success

Reply:
> Conversely, running regular mkdir on a directory that does exist also doesn't fail. Those both seem pretty broken to me.

i also challenge you on that. dfs_mkdir() will still fail.. again it depends on how you are handling the failure

Reply:
> I think `dfs_mkdir` returning EEXIST for both cases makes sense, so then `dfs_sys_mkdir_p` should handle this?

yes agreed. this is what my suggestion was.. we will **not** add an extra fetch on the happy path given we have conditional ops.

Reply:
> > Conversely, running regular mkdir on a directory that does exist also doesn't fail. Those both seem pretty broken to me.
> 
> i also challenge you on that. dfs_mkdir() will still fail.. again it depends on how you are handling the failure

I must have misunderstood what I was seeing. I just added a new unit test for the regular mkdir and it fails as expected.

Reply:
> > I think `dfs_mkdir` returning EEXIST for both cases makes sense, so then `dfs_sys_mkdir_p` should handle this?
> 
> yes agreed. this is what my suggestion was.. we will **not** add an extra fetch on the happy path given we have conditional ops.

OK, this makes sense. I think I'm going to need to do it for each parent, though. Sigh, it's never as easy as it first seems.